### PR TITLE
Gazelle: initial implementation of rule index

### PR DESCRIPTION
**Pull request by jayconrod**
(Unit tests will be in the PR to wire this into Resolver).

Related #859

### CODE REVIEW

File: go/tools/gazelle/config/constants.go [hunk 1]
```diff
@@ -49,3 +49,28 @@ const (
 	// on generated rules. It is replaced with "deps" during import resolution.
 	GazelleImportsKey = "_gazelle_imports"
 )
+
+// Language is the name of a programming langauge that Gazelle knows about.
+// This is used to specify import paths.
+type Language int
+
+const (
+	GoLang Language = iota
```
Comment (ianthehat) [inline go/tools/gazelle/config/constants.go]: Public symbols, trivial comments needed...

File: go/tools/gazelle/resolve/index.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+//
+// Any number of rules may provide the same import. If no rules provide
+// the import, ruleNotFoundError is returned. If multiple rules provide the
+// import, this function will attempt to choose one based on visibility.
+// An error is returned if the import is still ambiguous.
+//
+// Note that a rule may be returned even if visibility restrictions will be
+// be violated. Bazel will give a descriptive error message when a build
+// is attempted.
+func (ix *RuleIndex) findRuleByImport(imp importSpec, lang config.Language, fromRel string) (*ruleRecord, error) {
```
Comment (ianthehat) [inline go/tools/gazelle/resolve/index.go]: In the presence of multiple vendor folders, we may need to be able to pick the "nearest" in the hierarchy here as well, but it's not clear to me that there is enough information to do so in the structures.

File: go/tools/gazelle/resolve/index.go [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+		label = label.Abs("", buildRel)
+		if label.Repo != "" || label.Name != "__subpackages__" {
+			continue
+		}
+		visibility = append(visibility, label.Pkg)
+	}
+	return visibility
+}
+
+func isVisibleFrom(visibility visibilitySpec, defRel, useRel string) bool {
```
Comment (ianthehat) [inline go/tools/gazelle/resolve/index.go]: I am not convinced we should be handling visibility, I think it is going to be more surprising than helpful, especially as it's going to be hard to be sure we cover all the visibility cases.
I think the set of cases where visibility is going to help us pick the right one of multiple packages with the same import path are negligible to non existent, far more likely are the cases where we should generate the dep anyway so they get a nice "is not visible" error when they try to build the thing they really wanted.

File: go/tools/gazelle/resolve/index.go (hunk 2 above)
Comment (jayconrod) [reply go/tools/gazelle/resolve/index.go]: I've added this logic for Go only (not proto). I'm using the Pkg part of the label as a proxy for the directory in the repository. This won't work in flat mode or with manually written rules, but I'm not sure how much vendoring matters for those cases.

File: go/tools/gazelle/resolve/index.go (hunk 3 above)
Comment (jayconrod) [reply go/tools/gazelle/resolve/index.go]: You're right, I think I half-realized this when I allowed findRuleByImport to succeed when the only matches available were restricted by visibility.

I've removed all visibility parsing.

**Pull request merged**